src/multi/mod.rs

```diff
@@ -991,3 +998,194 @@ where
+              return Some(Err(Err::Error(E::append(input.clone(), ErrorKind::Many, e))));
+            } else {
+              return Some(Ok(input.clone()));
+            }
+          }
+          Err(e) => {
+            return Some(Err(e));
+          }
+        }
+    };
+    
+    match range.end_bound() {
+      Bound::Unbounded => for count in (1..=core::usize::MAX).map(|x| x - 1) {
+        match parser(count) {
+          Some(Ok(input)) => return Ok((input, res)),
+          Some(Err(e)) => return Err(e),
+          _ => {},
+        }
+      },
+      Bound::Included(x) => for count in (1..=*x).map(|x| x - 1) {
+        match parser(count) {
+          Some(Ok(input)) => return Ok((input, res)),
+          Some(Err(e)) => return Err(e),
+          _ => {},
+        }
+      },
+      Bound::Excluded(x) => for count in (1..*x).map(|x| x - 1) {
+        match parser(count) {
+          Some(Ok(input)) => return Ok((input, res)),
+          Some(Err(e)) => return Err(e),
+          _ => {},
+        }
+      },
+    }
+
+    Ok((input, res))
+  }
+}
+
+/// Applies a parser and accumulates the results using a given
+/// function and initial value.
+/// Fails if the amount of time the embedded parser is run is not
+/// within the specified range.
+/// 
+/// # Arguments
+/// * `range` The amount of times to apply the parser.
+///   * A range with an upper bound `a..=b` limits the parser to run at most `b` times.
+///   * A range without an upper bound `a..` allows the parser to run until it fails.
+///   * An empty range is invalid.
+/// * `parse` The parser to apply.
+/// * `init` A function returning the initial value.
+/// * `fold` The function that combines a result of `f` with
+///       the current accumulator.
+/// ```rust
+/// # #[macro_use] extern crate nom;
+/// # use nom::{Err, error::ErrorKind, Needed, IResult};
+/// use nom::multi::fold;
+/// use nom::bytes::complete::tag;
+///
+/// fn parser(s: &str) -> IResult<&str, Vec<&str>> {
+///   fold(
+///     0..=2,
+///     tag("abc"),
+///     Vec::new,
+///     |mut acc: Vec<_>, item| {
+///       acc.push(item);
+///       acc
+///     }
+///   )(s)
+/// }
+///
+/// assert_eq!(parser("abcabc"), Ok(("", vec!["abc", "abc"])));
+/// assert_eq!(parser("abc123"), Ok(("123", vec!["abc"])));
+/// assert_eq!(parser("123123"), Ok(("123123", vec![])));
+/// assert_eq!(parser(""), Ok(("", vec![])));
+/// assert_eq!(parser("abcabcabc"), Ok(("abc", vec!["abc", "abc"])));
+/// ```
+pub fn fold<I, O, E, F, G, H, J, R>(
+  range: J,
+  mut parse: F,
+  mut init: H,
+  mut fold: G,
+) -> impl FnMut(I) -> IResult<I, R, E>
+where
+  I: Clone + InputLength,
+  F: Parser<I, O, E>,
+  G: FnMut(R, O) -> R,
+  H: FnMut() -> R,
+  E: ParseError<I>,
+  J: NomRange<usize>,
+{
+  move |mut input: I| {
+    if range.is_inverted() {
+      return Err(Err::Failure(E::from_error_kind(input, ErrorKind::Fold)));
+    }
+
+    let mut acc = init();
+    
+    for count in range.saturating_iter().map(|x| x - 1) {
```

Comment:
again your `saturating_iter` don't make any sense and again you hide a ugly `SaturatingIter ` with it, and again a map that don't make any sense

Reply:
What is ugly about the Iterator now? Its definition is literally:
```rust
fn next(&mut self) -> Option<Self::Item> {
    let old_count = self.count;
    self.count = self.count.saturating_add(1);
    Some(old_count)
  }
```
It does not get more concise than that. And I cant just substitute it with a raw loop because we still have to count the number of successful parses because its later used to check if we fulfill the range (we could still be below the minimum and must therefore fail).

The exact type of `saturating_iter()` gets evaluated at compile time so there should be no additional overhead. What other problem could there be with it? The only alternative would be keeping count manually:

```rust
let mut count = 0;
loop {
  //parser stuff here
  count = count.saturating_add(1);
}
if (!range.contains(&count)) {
  //error here
```
Which I honestly think is more ugly and senseless than having a simple type that does it for you.

Reply:
it's becoming hard to follow since you name it the same that for your `RangeIterator` trait. And anyway, what is the point now ? use https://doc.rust-lang.org/std/primitive.u32.html#method.wrapping_add if you want to have it wrap nicely, check at every loop if this overflow is very very very not good and don't need this at all for unbound.

Reply:
Wrapping arithmetic is wrong and doesnt work.